### ISSUE TITLE
Using Timber library with Bedrock

### ISSUE THREAD

**Issue by chrisgherbert** [timber/timber#979]
I've been using Timber with [Bedrock](https://roots.io/bedrock/).

Until RC3, there were no major issues.  Now Timber calls `is_admin()` and uses WordPress hooks on being loaded.  Because Bedrock loads Composer dependencies before WordPress this will cause a fatal error.

Is the only solution to switch to the plugin version of Timber?

**Comment by connorjburton** [comment]
Hi, do you have stack traces please? Our preference is Timber is to be used with composer, so if we can solve this we will.

**Comment by chrisgherbert** [comment]
```
[05-May-2016 18:31:42 Europe/Berlin] PHP Fatal error:  Call to undefined function Timber\is_admin() in /Users/cherbert/web/bedrock-timber-test/vendor/timber/timber/lib/Timber.php on line 61
[05-May-2016 18:31:42 Europe/Berlin] PHP Stack trace:
[05-May-2016 18:31:42 Europe/Berlin] PHP   1. {main}() /Users/cherbert/web/bedrock-timber-test/web/index.php:0
[05-May-2016 18:31:42 Europe/Berlin] PHP   2. require() /Users/cherbert/web/bedrock-timber-test/web/index.php:5
[05-May-2016 18:31:42 Europe/Berlin] PHP   3. require_once() /Users/cherbert/web/bedrock-timber-test/web/wp/wp-blog-header.php:13
[05-May-2016 18:31:42 Europe/Berlin] PHP   4. require_once() /Users/cherbert/web/bedrock-timber-test/web/wp/wp-load.php:42
[05-May-2016 18:31:42 Europe/Berlin] PHP   5. require_once() /Users/cherbert/web/bedrock-timber-test/web/wp-config.php:7
[05-May-2016 18:31:42 Europe/Berlin] PHP   6. ComposerAutoloaderInit658bcd7a4d9b35310726249c356c62aa::getLoader() /Users/cherbert/web/bedrock-timber-test/vendor/autoload.php:7
[05-May-2016 18:31:42 Europe/Berlin] PHP   7. composerRequire658bcd7a4d9b35310726249c356c62aa() /Users/cherbert/web/bedrock-timber-test/vendor/composer/autoload_real.php:45
[05-May-2016 18:31:42 Europe/Berlin] PHP   8. require() /Users/cherbert/web/bedrock-timber-test/vendor/composer/autoload_real.php:55
[05-May-2016 18:31:42 Europe/Berlin] PHP   9. Timber\Timber->__construct() /Users/cherbert/web/bedrock-timber-test/vendor/timber/timber/init.php:5
[05-May-2016 18:31:42 Europe/Berlin] PHP  10. Timber\Timber->test_compatibility() /Users/cherbert/web/bedrock-timber-test/vendor/timber/timber/lib/Timber.php:49
```

**Comment by prisis** [comment]
And how you require Timbre in composer?

**Comment by chrisgherbert** [comment]
Here's the `require` object in my `composer.json`.  This is all standard Bedrock with the addition of Timber.

```
  "require": {
    "php": ">=5.5",
    "composer/installers": "~1.0.12",
    "vlucas/phpdotenv": "^2.0.1",
    "johnpbloch/wordpress": "4.5.1",
    "oscarotero/env": "^1.0",
    "roots/wp-password-bcrypt": "1.0.0",
    "timber/timber": "^1.0"
  },
```

**Comment by prisis** [comment]
Timbre need to change the type in composer.json to from library to wp-plugin so composer installiert can copy Timber to wp plugins

**Comment by chrisgherbert** [comment]
I'd rather not use it as a WordPress plugin, as I can't be sure that it's loaded before using it in my own WordPress plugins.  I usually have a plugin that sets up custom post types and extends TimberPost with subclasses tailored to each post type.

**Comment by connorjburton** [comment]
Yes `timber/timber` is supposed to be loaded via autoloader, not through a plugin. This will be fixed in `1.0.2` either way.

**Comment by connorjburton** [comment]
Hi, can you please try the fix in #997 (you should be able to do `require timber/timber:dev-cjb-autoload-root`)

**Comment by jarednova** [comment]
@chrisgherbert I merged @connorjburton's changes into master — can you give the newest version a test to confirm that it solves your issue?

**Comment by chrisgherbert** [comment]
I'm still having the same problem:

```
[16-May-2016 19:50:59 Europe/Berlin] PHP Fatal error:  Call to undefined function Timber\is_admin() in /Users/cherbert/web/testsite.org/vendor/timber/timber/lib/Timber.php on line 63
[16-May-2016 19:50:59 Europe/Berlin] PHP Stack trace:
[16-May-2016 19:50:59 Europe/Berlin] PHP   1. {main}() /Users/cherbert/web/testsite.org/public_html/index.php:0
[16-May-2016 19:50:59 Europe/Berlin] PHP   2. require() /Users/cherbert/web/testsite.org/public_html/index.php:5
[16-May-2016 19:50:59 Europe/Berlin] PHP   3. require_once() /Users/cherbert/web/testsite.org/public_html/wp/wp-blog-header.php:13
[16-May-2016 19:50:59 Europe/Berlin] PHP   4. require_once() /Users/cherbert/web/testsite.org/public_html/wp/wp-load.php:42
[16-May-2016 19:50:59 Europe/Berlin] PHP   5. require_once() /Users/cherbert/web/testsite.org/public_html/wp-config.php:7
[16-May-2016 19:50:59 Europe/Berlin] PHP   6. ComposerAutoloaderInite96a73f412326f44f8b2818fdbe77ead::getLoader() /Users/cherbert/web/testsite.org/vendor/autoload.php:7
[16-May-2016 19:50:59 Europe/Berlin] PHP   7. composerRequiree96a73f412326f44f8b2818fdbe77ead() /Users/cherbert/web/testsite.org/vendor/composer/autoload_real.php:45
[16-May-2016 19:50:59 Europe/Berlin] PHP   8. require() /Users/cherbert/web/testsite.org/vendor/composer/autoload_real.php:55
[16-May-2016 19:50:59 Europe/Berlin] PHP   9. Timber\Timber->__construct() /Users/cherbert/web/testsite.org/vendor/timber/timber/init.php:5
[16-May-2016 19:50:59 Europe/Berlin] PHP  10. Timber\Timber->test_compatibility() /Users/cherbert/web/testsite.org/vendor/timber/timber/lib/Timber.php:51
```

**Comment by connorjburton** [comment]
Hm, interesting, in theory it should work.

Is it possible to provide your code over email? My address is REDACTED - otherwise any debugging you can do on the new code in terms of checking if the conditional statement is running as expected would be great (Does `\WP` exist etc)

**Comment by chrisgherbert** [comment]
It's really just a standard Bedrock setup with Timber added as a Composer dependency.  Here's a repo I created that has the most recent commit of Timber as a dependency: https://github.com/chrisgherbert/bedrock-with-timber

If you're not familiar with installing a Bedrock project, you just need to do the following:

- Clone the repo
- Run `composer install`
- Copy `.env.example` to `.env`, then modify with DB credentials

That should get things going.

**Comment by joelambert** [comment]
Also having this issue, have tried the latest version in Packagist and also latest commit from `dev-master`.

Rolling back to `1.0-rc2` resolves the issue as a temporary fix.

**Comment by connorjburton** [comment]
Working on this, thanks for repo @chrisgherbert

**Comment by PhilippBaschke** [comment]
I am having the same problem and as far as I understand it has nothing to do with the `init` function. 

The files autoload
```json
  "autoload": {
    "files": ["init.php"]
  }
```
makes composer require `init.php` which creates a new `Timber` object and the constructor calls `test_compatibility` which calls the offending `is_admin`.

Hope that helps somehow (my problem might also be unrelated). :relaxed:

**Comment by connorjburton** [comment]
Ah okay, yeah I know what the issue is now. Thanks

**Comment by aaemnnosttv** [comment]
The `files` autoload is the problem and something to be really careful with since it loads at the same time as the composer's `autoload.php`.  In the case of Bedrock, it's loaded before WordPress is loaded at all, so any file that is being loaded there can't reference anything from WordPress because it won't exist yet.

Unfortunately, as a library it can't initialize itself without using something like the `files` autoload.  The alternative would be to use an mu-plugin to load/initialize the library (a possible companion package for project installs), but that also complicates things for non-bedrock projects, as that mu-plugin would still need to be loaded since `type: wordpress-muplugin` would be installed as a subdirectory and would not be loaded by default.  Bedrock has an autoloader which could do that automatically.

I think the best compromise may be - test if WordPress is loaded when the init is called, if not, return to prevent these kind of fatals, but allow for it to be called again manually.  That way, for those who are loading the package this way, all you need to do is call the init/bootstrap method yourself where it makes sense in your project.  The `timber.php` plugin file could call this after the code for the autoloader, making the `init.php` unnecessary.

I would be open to helping out here if you guys are interested?

**Comment by chrisgherbert** [comment]
How did this work before 1.0 RC3?

**Comment by connorjburton** [comment]
Hey guys, just committed, can you please test. I got the issue replicated on local, did the change and it seemed to resolve the error.

Thanks!

**Comment by chrisgherbert** [comment]
@connorjburton Thanks!  I tested the new commit and it doesn't immediate cause a fatal error.  However, when I try to activate the Timber Starter Theme I get the following:

`Timber not activated. Make sure you activate the plugin in /wp-admin/plugins.php`

**Comment by connorjburton** [comment]
If you go in your `functions.php` and remove the top conditional check to see if Timber exists and then try please?

**Comment by chrisgherbert** [comment]
Then I get a fatal error because the Timber class doesn't exist.

**Comment by connorjburton** [comment]
Okay, that's because `init.php` hasn't run. You'll have to manually init `Timber`, best bet here is to do the following in your `functions.php` I reckon.

```php
use Timber\Timber;
$timber = new Timber();
Timber::$dirname = 'views';
```

I would rather you not have to do that however, let me know if that works

**Comment by aaemnnosttv** [comment]
Couldn't that initialization be included into a single method call?

Why not just something like
```php
Timber\Timber::initialize();
```
then in Timber class

```php
public static function initialize()
{
    $GLOBALS['timber'] = new self();
    self::$dirname = 'views';
}
```

That could probably be improved a bit also to make it more flexible

```php
public static function initialize($views_directory = 'views')
{
    $GLOBALS['timber'] = new self();
    self::$dirname = $views_directory;
}
```

or maybe just set the default for that property in the declaration:

```php
class Timber {
// ..
    public static $dirname = 'views';
//..
}
```

**Comment by connorjburton** [comment]
The argument i'm having with myself is that with the new namespacing you shouldn't have to init Timber at all, as you can just use `use Timber\Timber` (in theory). The only reason you would have to do the above is if you are using composer/autoloading but not using the `use` command.

**Comment by chrisgherbert** [comment]
Even with `use Timber\Timber` it fails pretty quickly.  I think the start theme needs a pretty comprehensive renovation to support namespacing.

**Comment by mrgrain** [comment]
I would argue it is your responsibility as a theme builder to initialize Timber properly. I mean Timber is (now) supposed to be a library rather than a "working thing". Initialisation is quite common. One way to simplify this process, is to provide a `init.php` file which a theme might **choose** to include or not.

**Comment by chrisgherbert** [comment]
Fair enough, but the boilerplate Timber theme should probably provide that init.

**Comment by mrgrain** [comment]
Sure, it probably should. Are we talking about the [starter-theme](https://github.com/timber/starter-theme)? This is _technically_  speaking still based on [Timber 0.22](https://github.com/timber/starter-theme/blob/master/composer.json#L27), so I reckon it needs a major revamp anyway.

For Timber, it seems to be more an documentation issue.

**Comment by mrgrain** [comment]
hey @jarednova Could we please move the fix for this forward ASAP and release version 1.0.4? I think it makes a pretty bad impression if the preferred way of using/installing timber is completely broken.

However, it might be faster/easier/safer to exclude all the other changes and only release the fix. Let me know if I can help you with that in any way.

**Comment by jarednova** [comment]
@mrgrain thanks for the bug on this. I already pushed v1.0.4, but there's nothing stopping us from still prioritizing this and pushing as soon as it's ready given the importance of the issue.

So that I can fully understand: does #1041 close this comprehensively? Or would there be any other other parts of the issues described that would need a separate PR?

**Comment by mrgrain** [comment]
Hey @jarednova Thank you very much :bouquet: :smile: As mentioned elsewhere the additional check for WordPress in 1.0.4 is good enough to take out the pressure. We can move forward with 1.1.0 faster, but now we don't have to.
Yes, I think #1041 solves the composer use case comprehensively, but I might add a regression test for it over the weekend. However, in all honesty, I don't know Timber well enough to say whether this fixes every scenario. (Thinking in terms of multiple themes requiring timber installed, one of them maybe using the plugin version, etc.)
Now I'm happy to take some more time for 1.1.0 in case there is any other BC-breaking stuff we want to introduce.

**Comment by mrgrain** [comment]
@gridsystem No, it's unfortunately not documented anywhere right now. However in #835 the decision was made to position Timber as _a library_ rather than a plugin. (Since I was one of the advocates of this, shame on me for not updating the documentation!). That means, it's recommended that a theme bundles the Timber library, rather than relying on an installed plugin.

As a theme developer it would be smart to use Composer to manage your dependencies. But at the end, it's completely up to you. And once a theme is packaged and uploaded to wordpress.org, Composer is removed from the equation anyway.

The _plugin version_ of Timber is also still available via wordpress.org. Using wpackagist is fine, but after all simply a different way of installaling the _plugin version_ of Timber (which is not the recommended way).

However, (from my experience) a lot of themes never make their ways to wordpress.org since they are only developed for specific environments. In this case, a theme requiring timber as a library which finally used in a bedrock like set-up, composer will end up installing timber to the vendor directory outside the theme directory. And that's exactly where we want it to be. :)

**Comment by connorjburton** [comment]
Yes, I think `1.0.4` should resolve any immediate headaches, at least it's possible to require Timber from the root without it throwing a fatal error now, it is however not perfect and I should have considered requiring Timber from a location where WP is not yet active.

`1.1.0` will bring more sanity to this, I have a few ideas for `1.1.0` such as config setup (#1053) that is going to bring the whole init process together to a more cohesive and elegant process. I will also make sure the starter theme is up to date when we release `1.1.0`

Let's give this some thought and take our time to test in all use cases so we can put this issue to bed.

Thanks for everyone's help.

**Comment by gridsystem** [comment]
@mrgrain thanks for that! So remove the wp-packagist reference from composer.json, include it as a standard php lib, and crack on.

The only question I have is how do I then init timber from within my theme, which is what you have been discussing previously.

I have in functions.php

    $loader = include($_SERVER['DOCUMENT_ROOT'] . '/../vendor/autoload.php');
    Theme\Bootstrap::init();

Then bootstrap.php

    namespace Theme;
    use Timber\Timber;
    class Bootstrap {
    	public static function init() {
    		new \Timber\Site();
    		Assets::load();
    		Filters::add();
    	}
    }

But

    Strict Standards: Non-static method Timber\Timber::init() should not be called statically in [...]/vendor/timber/timber/lib/Timber.php on line 258

**Comment by connorjburton** [comment]
Whoops, that is a bug. It will try and auto-init, which is what it's attempting. I'll fix that

You can init Timber by doing `new Timber\Timber();` (assuming you don't have namespacing, which you do so `new Timber();` should work)

You can see an example here

https://github.com/timber/timber/blob/master/init.php

**Comment by gridsystem** [comment]
Boom. Got it. Thanks!

**Comment by gridsystem** [comment]
I have to say - I'm really happy with the direction you're taking with Timber. You're making WordPress quite enjoyable to work with.

**Comment by jarednova** [comment]
This should be resolved by #1041 — @gridsystem @chrisgherbert (or anyone else) please test agains the current `master` to see if the fix resolves this for you (thanks to @mrgrain for the PR!)
